File: ctf/src/main/java/com/efficios/jabberwocky/ctf/trace/CtfTraceIterator.java
/*
 * Copyright (C) 2017 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package com.efficios.jabberwocky.ctf.trace;

import java.util.NoSuchElementException;

import org.eclipse.jdt.annotation.Nullable;
import org.eclipse.tracecompass.ctf.core.CTFException;
import org.eclipse.tracecompass.ctf.core.event.IEventDefinition;
import org.eclipse.tracecompass.ctf.core.trace.CTFTraceReader;

import com.efficios.jabberwocky.ctf.trace.event.CtfTraceEvent;
import com.efficios.jabberwocky.trace.ITraceIterator;

public class CtfTraceIterator<E extends CtfTraceEvent> implements ITraceIterator<E> {

    private final CtfTrace<E> fOriginTrace;
    private final CTFTraceReader fTraceReader;

    private @Nullable IEventDefinition fCurrentEventDef;

    public CtfTraceIterator(CtfTrace<E> originTrace) {
        CTFTraceReader reader;
        try {
            reader = new CTFTraceReader(originTrace.getInnerTrace());
        } catch (CTFException e) {
            /*
             * If the CtfTrace was initialized successfully, creating an
             * iterator should not fail.
             */
            throw new IllegalStateException(e);
        }
        fOriginTrace = originTrace;
        fTraceReader = reader;
        fCurrentEventDef = reader.getCurrentEventDef();
    }

    @Override
    public boolean hasNext() {
        return (fCurrentEventDef != null);
    }

    @Override
    public E next() {
        IEventDefinition currentEventDef = fCurrentEventDef;
        if (currentEventDef == null) {
            throw new NoSuchElementException();
        }

        /* Wrap the current event into a JW event */
        E event = fOriginTrace.getEventFactory().createEvent(currentEventDef);

        /* Prepare the "next next" event */
        try {
            fTraceReader.advance();
            fCurrentEventDef = fTraceReader.getCurrentEventDef();
        } catch (CTFException e) {
            /* Shouldn't happen if we did the other checks correctly */
            throw new IllegalStateException(e);
        }
        return event;
    }

    @Override
    public void close() {
        fTraceReader.close();
    }

}

File: jabberwocky-ctf/src/main/kotlin/com/efficios/jabberwocky/ctf/trace/CtfTraceIterator.kt
/*
 * Copyright (C) 2017 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package com.efficios.jabberwocky.ctf.trace

import java.util.NoSuchElementException

import org.eclipse.tracecompass.ctf.core.CTFException
import org.eclipse.tracecompass.ctf.core.event.IEventDefinition
import org.eclipse.tracecompass.ctf.core.trace.CTFTraceReader

import com.efficios.jabberwocky.ctf.trace.event.CtfTraceEvent
import com.efficios.jabberwocky.trace.TraceIterator

open class CtfTraceIterator<out E : CtfTraceEvent>(private val originTrace: CtfTrace<E>) : TraceIterator<E> {

    private val traceReader: CTFTraceReader = try {
        CTFTraceReader(originTrace.innerTrace)
    } catch (e: CTFException) {
        /*
         * If the CtfTrace was initialized successfully, creating an
         * iterator should not fail.
         */
        throw IllegalStateException(e)
    }

    private var currentEventDef: IEventDefinition? = traceReader.currentEventDef

    override fun hasNext(): Boolean = (currentEventDef != null)

    override fun next(): E {
        val currentEventDef = currentEventDef ?: throw NoSuchElementException()

        /* Wrap the current event into a JW event */
        val event = originTrace.eventFactory.createEvent(currentEventDef)

        /* Prepare the "next next" event */
        try {
            traceReader.advance()
            this.currentEventDef = traceReader.getCurrentEventDef()
        } catch (e: CTFException) {
            /* Shouldn't happen if we did the other checks correctly */
            throw IllegalStateException(e)
        }

        return event
    }

    override fun close() {
        traceReader.close()
    }

}
